Avb and TA in-tree

jenswi-linaro opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jforissier commented on ta/mk/ta_dev_kit.mk:
```diff
@@ -79,19 +82,19 @@ clean:
+include  $(ta-dev-kit-dir$(sm))/mk/subdir.mk
 
-ifneq ($(binary),)
+ifneq ($(user-ta-uuid),)
```
I don't get it. Where is `$(user-ta-uuid)` defined? Why replace `$(binary)` with `$(user-ta-uuid)` only here?

jenswi-linaro replied:
I'll fix this in the rebase to follow

jforissier commented on ta/arch/arm/link.mk:
```diff
@@ -5,24 +5,24 @@ link-script-dep$(sm) = $(link-out-dir$(sm))/.ta.ld.d
-cleanfiles += $(link-out-dir$(sm))/$(binary).map
-cleanfiles += $(link-out-dir$(sm))/$(binary).stripped.elf
-cleanfiles += $(link-out-dir$(sm))/$(binary).ta
+all: $(link-out-dir$(sm))/$(user-ta-uuid).dmp \
```
Ah, now I get it ;-).
Still, it looks like we have a backward compatibility problem with the parent commit. Just a missing `user-ta-uuid := $(binary)` I would say. Or even better, do not introduce `$(user-ta-uuid)` in the parent commit.

jenswi-linaro replied:
True, I'll move the `user-ta-uuid` stuff into this commit.

jforissier commented on ta/mk/ta_dev_kit.mk:
```diff
@@ -19,7 +19,8 @@ endif
 
-binary := $(BINARY)
+user-ta-uuid := $(BINARY)
+user-ta-ldadd := $(LDADD)
```
Like `LDFLAGS` that you have removed in a previous commit, I think `LDADD` is not the proper name to be used here because it does not deal with cross-compilation properly. It has happened in the past that some distribution set some compiler or linker variables by default (CFLAGS, LDFLAGS...), for use when building host binaries, which then conflicts with our cross-build.

How about using `LDADD_$(sm)` instead? That would be `LDADD_ta_arm32` or `LDADD_ta_arm64`, similar to what we already do with `CFLAGS_` and `CROSS_COMPILE_`.

jenswi-linaro replied:
I was trying for a grace period for `BINARY` and `LDADD`. `LDADD` is used by at least on TA in optee_test.
I propose that we try to keep existing behavior with regards to `ta/mk/ta_dev_kit.mk` as much as possible in the PR.

Long term I'd like to replace what's in a TAs Makefile with what's in the TAs `user_ta.mk`, when we do that `LDADD` can be buried properly.

jforissier replied:
OK

jforissier commented on mk/compile.mk:
```diff
@@ -149,6 +149,11 @@ $(foreach f, $(srcs), $(eval $(call \
 $(foreach f, $(gen-srcs), $(eval $(call process_srcs,$(f),$$(basename $f).o)))
 
+# Handle specified source files, that is, files that have a specified path
+# but where the object file should to into a specified out directory
```
s/to/go/

jforissier commented on ta/mk/build-user-ta.mk:
```diff
@@ -0,0 +1,42 @@
+		$(default-user-ta-target)))
+
+ta-dev-kit-dir$(sm) := $(out-dir)/export-$(ta-target)
+# EARLY_TA_PATHS may hold paths to a TA in this directory
```
Does this comment relate to the next line?

jenswi-linaro replied:
It was only relevant in an earlier version of the patch set, I'll remove it.

jforissier commented on ta/avb/entry.c:
```diff
@@ -0,0 +1,258 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
C++-style comment for the SPDX line in a `.c` file

jenswi-linaro replied:
I'll fix

jforissier commented on ta/avb/entry.c:
```diff
@@ -0,0 +1,258 @@
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t __unused pt,
+				    TEE_Param __unused params[4],
+				    void __unused **session)
```
Move `__unused` after the parameter name, it will make checkpatch happy I suppose. It's what we have in all other optee_os files.

jenswi-linaro replied:
I'll fix

jenswi-linaro commented on ta/arch/arm/link.mk:
```diff
@@ -1,58 +1,62 @@
+link-ldflags += -Map=$(link-out-dir$(sm))/$(binary).map
 link-ldflags += --sort-section=alignment
 
-link-ldadd  = $(LDADD)
```
I'll fix.

jforissier commented:
One more (minor) thing: `include/private` rather than `private_include`?

jenswi-linaro commented:
Regarding `private_include`, I'd rather not put the content of the directory under `include` which is supposed to be exported in some way. Other than that I'd accept almost any name for the `private_include` directory.

jenswi-linaro commented:
Comments addressed and tags applied.
Note that the commit message of "Add support for compiling in-tree TAs" is slightly different from what was requested. Also `LDADD` is still used when compiling TAs via `ta-dev-kit.sh`

jforissier commented:
- Why not avoid `private_include` altogether, and put `user_ta_header_defines.h` directly under `avb`? That would be my preferred choice, otherwise let's keep `private_include`.

- The commit message of "Add support for compiling in-tree TAs" is OK, TBH I had not taken a close enough look at what it was doing exactly ;-)

- I'm still a bit concerned that commit "ta: link.mk: drop support for variable LDFLAGS"  [note: should add "and LDADD" BTW] can needlessly break bisectability, since you're re-introducing the variables in a later commit. Not a big deal anyway.

- Commits "ta: prepare for in-tree ta building" and "plat-hikey: embed ta/avb as early TA" are
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
(forgot to mention them above, sorry for that).

- For "ta: support variables user-ta-{uuid,ldadd}":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`.

jenswi-linaro commented:
Tags applied
Restored `LDADD` in "ta: link.mk: drop support for variable LDFLAGS"
Dropped `private_include` and moved `user_ta_header_defines.h` directly under `avb`

jenswi-linaro commented:
Addressed the last checkpatch error

jenswi-linaro commented:
Addressed the last checkpatch error in "ta: support variables user-ta-{uuid,ldadd}"
